src/main/java/gyro/google/compute/FirewallRuleResource.java

```diff
@@ -0,0 +1,580 @@
+        setSourceTags(firewall.getSourceTags() != null ? new HashSet<>(firewall.getSourceTags()) : null);
+        setLogConfig(firewall.getLogConfig().getEnable());
+        setTargetServiceAccounts(firewall.getTargetServiceAccounts() != null ? new HashSet<>(firewall.getTargetServiceAccounts()) : null);
+        setTargetTags(firewall.getTargetTags() != null ? new HashSet<>(firewall.getTargetTags()) : null);
+
+        getRule().clear();
+        if (firewall.getAllowed() != null && !firewall.getAllowed().isEmpty()) {
+            setRule(firewall.getAllowed().stream().map(rule -> {
+                FirewallAllowDenyRule allowedRule = new FirewallAllowDenyRule();
+                allowedRule.copyFrom(rule);
+                return allowedRule;
+            }).collect(Collectors.toSet()));
+        } else if (firewall.getDenied() != null && !firewall.getDenied().isEmpty()) {
+            setRule(firewall.getDenied().stream().map(rule -> {
+                FirewallAllowDenyRule deniedRule = new FirewallAllowDenyRule();
+                deniedRule.copyFrom(rule);
+                return deniedRule;
+            }).collect(Collectors.toSet()));
+        }
+
+        setId(firewall.getId().toString());
+        setSelfLink(firewall.getSelfLink());
+    }
+
+    @Override
+    public boolean refresh() {
+        Compute client = creatClient(Compute.class);
+
+        Firewall firewall = getFirewall(client);
+
+        if (firewall == null) {
+            return false;
+        }
+
+        copyFrom(firewall);
+
+        return true;
+    }
+
+    @Override
+    public void create(GyroUI ui, State state) {
+        Compute client = creatClient(Compute.class);
+
+        try {
+            Compute.Firewalls.Insert insert = client.firewalls().insert(getProjectId(), toFirewall(client, ACTION_TYPE.CREATE));
+            Operation operation = insert.execute();
+            Operation.Error error = waitForCompletion(client, operation);
+            if (error != null) {
+                throw new GyroException(error.toPrettyString());
+            }
+
+            Firewall firewall = getFirewall(client);
+            setId(firewall.getId().toString());
+            setSelfLink(firewall.getSelfLink());
+        } catch (Exception ex) {
+            throw new GyroException(ex.getMessage(), ex.getCause());
+        }
+    }
+
+    @Override
+    public void update(GyroUI ui, State state, Resource current, Set<String> changedFieldNames) {
+        Compute client = creatClient(Compute.class);
+
+        Operation operation;
+        Operation.Error error;
+
+        try {
+            // update
+            operation = client.firewalls().update(getProjectId(), getName(), toFirewall(client, ACTION_TYPE.UPDATE)).execute();
+            error = waitForCompletion(client, operation);
+            if (error != null) {
+                throw new GyroException(error.toPrettyString());
+            }
+
+            // patch
```

Comment:
I'm not sure I follow the need for using both update and patch. Isn't patch just an optimized variant of update?

https://cloud.google.com/compute/docs/api/how-tos/performance#partial

Reply:
The update api allows for certain fields to be updated, for the rest the patch needs to be used. As per their docs they specify the ones `update` supports:
```
/**
     * Updates the specified firewall rule with the data included in the request. The PUT method can
     * only update the following fields of firewall rule: allowed, description, sourceRanges,
     * sourceTags, targetTags.
     *
     *
```

Reply:
We figured that patch is able to update all the fields and so the `gyro update` now uses patch to update all the fields for the firewall rule.